crates/runtime/src/lib.rs

```diff
@@ -1,13 +1,223 @@
-mod borrow;
+use std::cell::Cell;
+use std::fmt;
+use std::marker;
+use std::slice;
+use std::str;
+
 mod error;
 mod guest_type;
-mod memory;
 mod region;
-
 pub use error::GuestError;
-pub use guest_type::{GuestErrorType, GuestType, GuestTypeTransparent};
-pub use memory::{
-    GuestArray, GuestMemory, GuestPtr, GuestPtrMut, GuestRef, GuestRefMut, GuestString,
-    GuestStringRef,
-};
+pub use guest_type::{GuestErrorType, GuestType};
 pub use region::Region;
+
+pub unsafe trait GuestMemory {
+    fn base(&self) -> (*mut u8, u32);
+
+    fn validate_size_align(
+        &self,
+        offset: u32,
+        align: usize,
+        len: u32,
+    ) -> Result<*mut u8, GuestError> {
+        let (base_ptr, base_len) = self.base();
+        let region = Region { start: offset, len };
```

Comment:
Oh no that's true, `Region` isn't used any more because "borrow checking" is left to users rather than implemented as part of the runtime. I don't mind removing it here, this was just the first piece of code I wrote and wasn't sure whether `Region` would survive all the way to the end (turns out it hasn't).

I've been pretty sloppy with errors in a few other places, so once this "generally looks good" I'll put some work into polishing it.

Reply:
Yep, thought as much, thanks for clarifying! I'd leave it as-is in this PR and refactor things out after this PR lands :-)